D103-D106 QDYNAM instrumentation (QUS3D.f part: the live moisture-advection routines AADVQ/AADVQ0/aadvqx/y/z/checkflux live in QUS3D.f, not QUS_DRV.f; hooks ffde_stg/ffde_carry/ffde_full inside AADVQ). Helpers are in ATM_DRV_dynE.f.patch. Generated by diff -u against the PRISTINE QUS3D.f (2026-10-05); local hunks.

--- QUS3D.f.orig
+++ QUS3D.f
@@ -207,12 +207,18 @@
      &                   rm(i,j,l),rmom(my,i,j,l),rmom(myy,i,j,l))
                   enddo
                 enddo
+                call ffde_stg(0,nc,ncxy,l,RM(1,j_0h,l),
+     &               RMOM(1,1,j_0h,l),MMA(1,j_0h,l))
                 CALL AADVQY(RM(1,j_0h,l),RMOM(1,1,j_0h,l),
      &               MMA(1,j_0h,l),pv(1,j_0h,l)
      &               ,sbf(j_0h,l),sbm(j_0h,l),sfbm(j_0h,l),sbfv)
+                call ffde_stg(1,nc,ncxy,l,RM(1,j_0h,l),
+     &               RMOM(1,1,j_0h,l),MMA(1,j_0h,l))
                 CALL AADVQX(RM(1,j_0h,l),RMOM(1,1,j_0h,l),
      &               MMA(1,j_0h,l),pu(1,j_0h,l),jmin_x,jmax_x,
      &               nstepx(j_0h,l),safv)
+                call ffde_stg(2,nc,ncxy,l,RM(1,j_0h,l),
+     &               RMOM(1,1,j_0h,l),MMA(1,j_0h,l))
               else
                 CALL AADVQY2(RM(1,j_0h,l),RMOM(1,1,j_0h,l),
      &               MMA(1,j_0h,l),pv(1,j_0h,l)
@@ -235,6 +241,9 @@
               enddo
             enddo
           endif
+          if(qlimit .and. l.gt.1 .and. l.lt.lm)
+     &       call ffde_stg(3,nc,0,l,RM(1,j_0h,l),
+     &       RMOM(1,1,j_0h,l),MMA(1,j_0h,l))
           if(l.gt.1) then
             if(qlimit) then
               CALL AADVQZ(RM(1,j_0h,l-1),RMOM(1,1,j_0h,l-1),
@@ -242,6 +251,7 @@
      &             ,scf(j_0h,l-1),scm(j_0h,l-1),sfcm(j_0h,l-1))
               SCF3D(:,j_0:j_1,L-1) = SCF3D(:,j_0:j_1,L-1) + FDN(:
      &             ,j_0:j_1)
+              call ffde_carry(nc,l,mwdn,fdn,fdn0,fmomdn)
             else
               CALL AADVQZ2(RM(1,j_0h,l-1),RMOM(1,1,j_0h,l-1),
      &             MMA(1,j_0h,l-1),SD(1,j_0h,L-1),mwdn,fdn,fmomdn!,fdn0
@@ -252,6 +262,7 @@
           endif
 
         ENDDO                   ! l
+        call ffde_full(6,nc,rm,rmom,MMA)
 
 c
 c perform the extra z advection "against the grain"
@@ -282,6 +293,7 @@
           enddo ! i
           enddo ! j
         endif ! do_z_extra
+        if(do_z_extra) call ffde_full(7,nc,rm,rmom,MMA)
 
       enddo                     ! ncyc
 
